### PR TITLE
Basic user manage his badges / user not  mandatory in tags anymore

### PR DESCRIPTION
**Pull request by mahdibenromdhane**

### CODE REVIEW

File: src/server/rest/v1/service/TagService.ts [hunk 1]
```diff
@@ -66,6 +65,35 @@ export default class TagService {
     next();
   }
 
+  public static async handleUnassignTags(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
+    // Filter
+    const visualTagsIDs = TagValidator.getInstance().validateTagsUnassign(req.body).visualTagsIDs;
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:70]: visualIDs

File: src/server/rest/v1/service/TagService.ts [hunk 2]
```diff
@@ -182,10 +346,10 @@ export default class TagService {
...
     const tag = await UtilsService.checkAndGetTagAuthorization(req.tenant, req.user, filteredRequest.id, Action.UPDATE, action,
       filteredRequest, { withNbrTransactions: true, withUser: true }, true);
-    // Get User
-    const user = await UtilsService.checkAndGetUserAuthorization(req.tenant, req.user, filteredRequest.userID,
-      Action.READ, ServerAction.TAG_UPDATE);
-    // Check visualID uniqueness
+    if (filteredRequest.userID) {
+      await UtilsService.checkAndGetUserAuthorization(req.tenant, req.user, filteredRequest.userID,
+        Action.READ, ServerAction.TAG_UPDATE);
+    } // Check visualID uniqueness
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:352]: Comment in the next line

File: src/server/rest/v1/service/TagService.ts [hunk 3]
```diff
@@ -217,17 +381,17 @@ export default class TagService {
       formerTagDefault = tag.default;
     }
     // Clear User's default Tag
-    if (filteredRequest.default && !formerTagUserID && (tag.default !== filteredRequest.default)) {
+    if (filteredRequest.default && !formerTagUserID && filteredRequest.userID && (tag.default !== filteredRequest.default)) {
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:384]: Put line 384 to 396 into the if (filteredRequest.userID) {...}

File: src/server/rest/v1/service/TagService.ts [hunk 4]
```diff
@@ -547,6 +712,51 @@ export default class TagService {
     return result;
   }
 
+  private static async unassignTags(tenant: Tenant, action: ServerAction, loggedUser: UserToken, visualTagsIDs: string[]): Promise<ActionsResponse> {
+    const result: ActionsResponse = {
+      inSuccess: 0,
+      inError: 0
+    };
+    // Delete Tags
+    for (const visualTagsID of visualTagsIDs) {
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:721]: visualIDs

File: src/server/rest/v1/service/TagService.ts [hunk 5]
```diff
@@ -547,6 +712,51 @@ export default class TagService {
...
+      try {
+        let tag = await TagStorage.getTagByVisualID(tenant, visualTagsID);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:723]: Remove get of tag as it is already done in 725 with method checkAndGetTagAuthorization()

File: src/server/rest/v1/validator/TagValidator.ts [hunk 6]
```diff
@@ -10,19 +10,29 @@ export default class TagValidator extends SchemaValidator {
...
   private constructor() {
     super('TagValidator');
     this.importedTagCreation = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/imported-tag-create-req.json`, 'utf8'));
     this.tagCreate = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-create.json`, 'utf8'));
+    this.tagAssign = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-assign.json`, 'utf8'));
     this.tagUpdate = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-update.json`, 'utf8'));
+    this.tagUpdateByVisualID = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-update-by-visual-id.json`, 'utf8'));
     this.tagsGet = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tags-get.json`, 'utf8'));
     this.tagGet = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-get.json`, 'utf8'));
+    this.tagGetByVisualID = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-get-by-visual-id.json`, 'utf8'));
```
Comment (LucasBrazi06) [inline src/server/rest/v1/validator/TagValidator.ts:32]: tagGetByVisualID -> tagGet (always use visualID)

File: src/server/rest/v1/validator/TagValidator.ts [hunk 7]
```diff
@@ -10,19 +10,29 @@ export default class TagValidator extends SchemaValidator {
...
+  private tagsUnassign: Schema;
+  private tagUnassign: Schema;
 
   private constructor() {
     super('TagValidator');
     this.importedTagCreation = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/imported-tag-create-req.json`, 'utf8'));
     this.tagCreate = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-create.json`, 'utf8'));
+    this.tagAssign = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-assign.json`, 'utf8'));
     this.tagUpdate = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-update.json`, 'utf8'));
+    this.tagUpdateByVisualID = JSON.parse(fs.readFileSync(`${global.appRoot}/assets/server/rest/v1/schemas/tag/tag-update-by-visual-id.json`, 'utf8'));
```
Comment (LucasBrazi06) [inline src/server/rest/v1/validator/TagValidator.ts:29]: tagUpdateByVisualID -> tagUpdate (always use visualID)

File: src/server/rest/v1/validator/TagValidator.ts [hunk 8]
```diff
@@ -41,11 +51,21 @@ export default class TagValidator extends SchemaValidator {
...
+    this.validate(this.tagAssign, tag);
+    return tag;
+  }
+
   validateTagUpdate(tag: Tag): Tag {
     this.validate(this.tagUpdate, tag);
     return tag;
   }
 
+  validateTagUpdateByVisualID(tag: Partial<Tag>): Partial<Tag> {
```
Comment (LucasBrazi06) [inline src/server/rest/v1/validator/TagValidator.ts:64]: Should return Tag and not Partial<Tag>

File: src/types/GlobalType.ts [hunk 9]
```diff
@@ -48,7 +48,7 @@ export enum ImportStatus {
   ERROR = 'E',
 }
 
-export type EntityDataType = Car|User|Company|Site|SiteArea|Tag|CarCatalog|ChargingStation;
+export type EntityDataType = Car|User|Company|Site|SiteArea|Tag|CarCatalog|ChargingStation|Partial<Tag>;
```
Comment (LucasBrazi06) [inline src/types/GlobalType.ts:51]: Partial<Tag> -> Tag

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
Just a small remark
I'll review the front-end now.

File: src/server/rest/v1/service/TagService.ts [hunk 10]
```diff
@@ -66,6 +65,35 @@ export default class TagService {
     next();
   }
 
+  public static async handleUnassignTags(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
+    // Filter
+    const visualIDs = TagValidator.getInstance().validateTagsUnassign(req.body).visualIDs;
```
[the remark this answers is not in the document]
Comment (mahdibenromdhane) [reply src/server/rest/v1/service/TagService.ts]: Ok :) , Done

File: src/assets/server/rest/v1/schemas/tag/tag-assign.json [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  "description": "",
+  "type": "object",
+  "properties": {
+    "visualID": {
+      "$ref": "tag.json#/definitions/visualID"
+    },
+    "description": {
+      "$ref": "tag.json#/definitions/description"
+    },
+    "active": {
```
Comment (LucasBrazi06) [inline src/assets/server/rest/v1/schemas/tag/tag-assign.json:12]: Should not be part of the Assign request

File: src/assets/server/rest/v1/schemas/tag/tag-assign.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      "$ref": "tag.json#/definitions/active"
+    },
+    "issuer": {
```
Comment (LucasBrazi06) [inline src/assets/server/rest/v1/schemas/tag/tag-assign.json:15]: Should not be part of the Assign request

File: src/assets/server/rest/v1/schemas/tag/tag-assign.json (hunk 11 above)
Comment (mahdibenromdhane) [reply src/assets/server/rest/v1/schemas/tag/tag-assign.json]: Done

File: src/assets/server/rest/v1/schemas/tag/tag-assign.json (hunk 12 above)
Comment (mahdibenromdhane) [reply src/assets/server/rest/v1/schemas/tag/tag-assign.json]: Done

File: src/server/rest/v1/service/TagService.ts [hunk 13]
```diff
@@ -174,6 +204,145 @@ export default class TagService {
     next();
   }
 
+  public static async handleAssignTag(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
+    const filteredRequest = TagValidator.getInstance().validateTagAssign(req.body);
+    // Get dynamic auth
+    const authorizationFilter = await AuthorizationService.checkAndGetTagAuthorizations(
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TagService.ts:210]: Why don't you use:
```
const tag = await UtilsService.checkAndGetTagByVisualIDAuthorization(req.tenant, req.user, filteredRequest.visualID, Action.UPDATE_BY_VISUAL_ID, action,
      filteredRequest, { withNbrTransactions: true, withUser: true });
```

File: src/authorization/AuthorizationsDefinition.ts [hunk 14]
```diff
@@ -532,12 +532,37 @@ const AUTHORIZATION_DEFINITION: AuthorizationDefinition = {
...
+          'userID', 'issuer', 'active', 'description', 'visualID', 'default',
           'user.id', 'user.name', 'user.firstName', 'user.email', 'user.issuer'
         ],
       },
       {
-        resource: Entity.TAG, action: [Action.DELETE, Action.UPDATE],
+        resource: Entity.TAGS, action: Action.UNASSIGN,
+        condition: {
+          Fn: 'custom:dynamicAuthorizations',
+          args: { filters: ['OwnUser'] }
```
Comment (LucasBrazi06) [inline src/authorization/AuthorizationsDefinition.ts:543]: I should have 'assert' here to ensure that the Basic user can only change his own badge.

File: src/authorization/AuthorizationsDefinition.ts [hunk 15]
```diff
@@ -532,12 +532,37 @@ const AUTHORIZATION_DEFINITION: AuthorizationDefinition = {
...
+        }
+      },
+      {
+        resource: Entity.TAG, action: Action.UNASSIGN,
+        condition: {
+          Fn: 'custom:dynamicAuthorizations',
+          args: { filters: ['OwnUser'] }
```
Comment (LucasBrazi06) [inline src/authorization/AuthorizationsDefinition.ts:550]: Same here

File: src/authorization/AuthorizationsDefinition.ts [hunk 16]
```diff
@@ -532,12 +532,37 @@ const AUTHORIZATION_DEFINITION: AuthorizationDefinition = {
...
+        }
+      },
+      {
+        resource: Entity.TAG, action:  Action.UPDATE_BY_VISUAL_ID,
+        condition: {
+          Fn: 'custom:dynamicAuthorizations',
+          args: { filters: ['OwnUser'] }
```
Comment (LucasBrazi06) [inline src/authorization/AuthorizationsDefinition.ts:557]: Same here

File: src/authorization/AuthorizationsDefinition.ts [hunk 17]
```diff
@@ -532,12 +532,37 @@ const AUTHORIZATION_DEFINITION: AuthorizationDefinition = {
...
+        }
+      },
+      {
+        resource: Entity.TAG, action: Action.ASSIGN,
+        attributes: [
+          'userID', 'active', 'ocpiToken', 'description', 'visualID', 'issuer', 'default',
+          'user.name', 'user.firstName', 'user.email', 'createdOn', 'lastChangedOn'
+        ],
```
Comment (LucasBrazi06) [inline src/authorization/AuthorizationsDefinition.ts:565]: Should not have access to some fields: active, ocpiToken...

File: src/authorization/AuthorizationsDefinition.ts [hunk 18]
```diff
@@ -532,12 +532,37 @@ const AUTHORIZATION_DEFINITION: AuthorizationDefinition = {
...
         condition: {
           Fn: 'custom:dynamicAuthorizations',
           args: { filters: ['OwnUser'] }
```
Comment (LucasBrazi06) [inline src/authorization/AuthorizationsDefinition.ts:568]: Same here

File: src/types/Authorization.ts [hunk 19]
```diff
@@ -183,7 +184,10 @@ export interface AuthorizationActions {
   canRead?: boolean;
   canCreate?: boolean;
   canUpdate?: boolean;
+  canUpdateByVisualID?: boolean;
   canDelete?: boolean;
+  canUnassign?: boolean;
+  canAssign?: boolean;
```
Comment (LucasBrazi06) [inline src/types/Authorization.ts:190]: Create a new type TagAuthorizationActions and move the 3 canXXX inside.

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/types/Authorization.ts (hunk 19 above)
Comment (mahdibenromdhane) [reply src/types/Authorization.ts]: Done :)

**Review by LucasBrazi06 — APPROVED** [review]

**Pull request merged**